hack/generator/pkg/astbuilder/builder.go

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package astbuilder
+
+import (
+	"fmt"
+	"go/ast"
+	"go/token"
+)
+
+// CheckErrorAndReturn checks if the err is non-nil, and if it is returns.
+func CheckErrorAndReturn(otherReturns ...*ast.Ident) ast.Stmt {
+
+	var returnValues []ast.Expr
+	for _, ret := range otherReturns {
+		returnValues = append(returnValues, ret)
+	}
+	returnValues = append(returnValues, ast.NewIdent("err"))
+
+	return &ast.IfStmt{
+		Cond: &ast.BinaryExpr{
+			X:  ast.NewIdent("err"),
+			Op: token.NEQ,
+			Y:  ast.NewIdent("nil"),
+		},
+		Body: &ast.BlockStmt{
+			List: []ast.Stmt{
+				&ast.ReturnStmt{
+					Results: returnValues,
+				},
+			},
+		},
+	}
+}
+
+// NewStruct creates a new assignment statement where a struct is constructed and stored in a variable of the given name
+func NewStruct(varName *ast.Ident, structName *ast.Ident) ast.Stmt {
+	return &ast.AssignStmt{
+		Lhs: []ast.Expr{
+			varName,
+		},
+		Tok: token.DEFINE,
+		Rhs: []ast.Expr{
+			&ast.CompositeLit{
+				Type: structName,
+			},
+		},
+	}
+}
+
+// Check if a variable is nil and if it is, return
+func NilGuardAndReturn(toCheck *ast.Ident, returns ...*ast.Ident) ast.Stmt {
+
+	var returnValues []ast.Expr
+	for _, ret := range returns {
+		returnValues = append(returnValues, ret)
+	}
+
+	return &ast.IfStmt{
+		Cond: &ast.BinaryExpr{
+			X:  toCheck,
+			Op: token.EQL,
+			Y:  ast.NewIdent("nil"),
+		},
+		Body: &ast.BlockStmt{
+			List: []ast.Stmt{
+				&ast.ReturnStmt{
+					Results: returnValues,
+				},
+			},
+		},
+	}
+}
+
+// DefineFunc defines a function (header, body, etc)
+func DefineFunc(
+	receiverIdent *ast.Ident,
+	receiverType ast.Expr,
+	name *ast.Ident,
+	comment string,
+	params []*ast.Field,
+	returns []*ast.Field,
+	body []ast.Stmt) *ast.FuncDecl {
+
+	// TODO: I'm not actually sure this is better than just declaring it all inline...?
+	// TODO: Maybe if this took a struct param instead of so many individual items?
+
+	return &ast.FuncDecl{
+		Recv: &ast.FieldList{
+			List: []*ast.Field{
+				{
+					Names: []*ast.Ident{
+						receiverIdent,
+					},
+					Type: receiverType,
+				},
+			},
+		},
+		Name: name,
+		Doc: &ast.CommentGroup{
+			List: []*ast.Comment{
+				{
+					Text: fmt.Sprintf("// %s %s", name.Name, comment),
+				},
+			},
+		},
+		Type: &ast.FuncType{
+			Params: &ast.FieldList{
+				List: params,
+			},
+			Results: &ast.FieldList{
+				List: returns,
+			},
+		},
+		Body: &ast.BlockStmt{
+			List: body,
+		},
+	}
+}
+
+// SimpleAssignment performs a simple assignment
+func SimpleAssignment(lhs ast.Expr, tok token.Token, rhs ast.Expr) *ast.AssignStmt {
```

Comment:
Why does this need `tok`? Won't it always be assignment (by definition)?

Reply:
Not necessarily. This supports:
`<lhs> = <rhs>` and `<lhs> := <rhs>`

Possibly the language (assignment) is throwing a wrench here - but that's what the go ast calls the statement (and it supports both "definition" and "assignment"). I was using that terminology here.

If you're getting at the fact that this is _too_ flexible (because a tok could be anything), I sorta agree although again the ast lib doesn't make that distinction and allows asts that don't actually compile like assignment statement `foo } bar`. We could throw a `panic` in if `tok` is anything other than `DEFINE` or `ASSIGN`

Reply:
What about splitting this into two methods, one for each legit method:
``` go
func SimpleAssignment(lhs ast.Expr, rhs ast.Expr) *ast.AssignStmt {
```
and
```
func DeclarationAssignment(lhs ast.Expr, rhs ast.Expr) *ast.AssignStmt {
```
Thoughts?

Reply:
I thought about this and initially liked it, but then realized there are actually quite a few tokens which are valid here:
`+=`, `-=`, `&=`, `=`, `:=`, etc -- there are a total of 13 which will work so I think actually we should keep it as is.

I'm open to changing the name of the function though to be more generic (although I am not really sure what we can change it to) - the only thing I came up with was `SimpleAssignmentStatement` but I didn't really like that because none of the other helpers really refer to the "ast bits" directly (i.e. it's `NilGuardAndReturn` not `NilGuardAndReturnStatement`)